web/libs/editor/src/components/KonvaVector/components/ControlPoints.tsx

```diff
@@ -0,0 +1,107 @@
+import React from "react";
+import { Line, Rect } from "react-konva";
+import type { BezierPoint } from "../types";
+
+interface ControlPointsProps {
+  initialPoints: BezierPoint[];
+  selectedPointIndex: number | null;
+  isDraggingNewBezier: boolean;
+  draggedControlPoint: { pointIndex: number; controlIndex: number } | null;
+  visibleControlPoints: Set<number>;
+  transform: { zoom: number; offsetX: number; offsetY: number };
+  fitScale: number;
+}
+
+export const ControlPoints: React.FC<ControlPointsProps> = ({
+  initialPoints,
+  selectedPointIndex,
+  isDraggingNewBezier,
+  draggedControlPoint,
+  visibleControlPoints,
+  transform,
+  fitScale,
+}) => {
+  return (
+    <>
+      {initialPoints.map((point, index) => {
+        if (!point.isBezier) return null;
+
+        // Show control points if point is selected, we're creating this specific Bezier point, control points are marked as visible, or it's a bezier point
+        const shouldShowControls =
+          selectedPointIndex === index ||
+          (isDraggingNewBezier && draggedControlPoint?.pointIndex === index) ||
+          visibleControlPoints.has(index) ||
+          point.isBezier;
+        if (!shouldShowControls) return null;
+
+        // Scale up radius to compensate for Layer scaling
+        const scale = transform.zoom * fitScale;
+        const scaledRadius = 4 / scale;
+
+        return (
+          <React.Fragment
+            key={`control-group-${index}-${point.x.toFixed(2)}-${point.y.toFixed(2)}-${point.controlPoint1?.x.toFixed(2) || "null"}-${point.controlPoint1?.y.toFixed(2) || "null"}-${point.controlPoint2?.x.toFixed(2) || "null"}-${point.controlPoint2?.y.toFixed(2) || "null"}`}
```

Comment:
maybe we should memoize this string so it doesn't get rebuilt on every re-render

Reply:
it will still change every time point's moved. shouldn't be an issue